tap_mixpanel/streams.py

```diff
@@ -1,125 +1,654 @@
+                to_date = "{}".format(end_window.astimezone(tzone))[0:10]
+                LOGGER.info(
+                    "START Sync for Stream: {}{}".format(
+                        self.tap_stream_id,
+                        ", Date window from: {} to {}".format(
+                            from_date, to_date)
+                        if self.bookmark_query_field_from
+                        else "",
+                    )
+                )
+                params[self.bookmark_query_field_from] = from_date
+                params[self.bookmark_query_field_to] = to_date
+
+            # funnels and cohorts have a parent endpoint with parent_data and parent_id_field
+            if self.parent_path and self.parent_id_field:
+                # API request data
+                LOGGER.info(
+                    "URL for Parent Stream {}: {}/{}".format(
+                        self.tap_stream_id, self.url, self.parent_path
+                    )
+                )
+                parent_data = self.client.request(
+                    method="GET",
+                    url=self.url,
+                    path=self.parent_path,
+                    endpoint="parent_data",
+                )
+            # Other endpoints (not funnels, cohorts): Simulate parent_data with single record
+            else:
+                parent_data = [{"id": "none"}]
+                self.parent_id_field = "id"
+
+            for parent_record in parent_data:
+                parent_id = parent_record.get(self.parent_id_field)
+                LOGGER.info('START: Stream: {}, parent_id: {}'.format(
+                    self.tap_stream_id, parent_id))
+
+                # pagination: loop thru all pages of data using next (if not None)
+                page = 0  # First page is page=0, second page is page=1, ...
+                offset = 0
+                limit = 250  # Default page_size
+                # Initialize counters
+                parent_total = 0  # Total records for parent ID
+                total_records = 0  # Total records for all pages
+                record_count = 0  # Total processed for page
+
+                session_id = 'initial'
+                if self.pagination:
+                    params['page_size'] = limit
+
+                # Poped session_id and page number of last parents stream call.
+                params.pop('session_id', None)
+                params.pop('page', None)
+
+                while offset <= total_records and session_id is not None:
+                    if self.pagination and page != 0:
+                        params['session_id'] = session_id
+                        params['page'] = page
+
+                    # querystring: Squash query params into string and replace [parent_id]
+                    querystring = '&'.join(['%s=%s' % (key, value) for (key, value)
+                                            in params.items()]).replace(
+                        '[parent_id]', str(parent_id))
```

Comment:
I think the spacing here is hard to read. Breaking it up might make it better

```suggestion
                    querystring = '&'.join(['%s=%s' % (key, value) for (key, value) in params.items()])
                    querystring = querystring.replace('[parent_id]', str(parent_id))
```

Reply:
Updated